sdk/storage/Azure.Storage.Common.DataMovement/src/FileSystemScanner.cs

```diff
@@ -0,0 +1,116 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Azure.Storage.Common.DataMovement
+{
+    internal class FileSystemScanner
+    {
+        /// <summary>
+        /// The fully qualified path to be scanned.
+        /// </summary>
+        public string _basePath;
+
+        private DirectoryScannerFactory _dirFactory;
+        private DirectoryScanner _currentDirectory;
+
+        private Queue<string> _directories;
+        private Queue<string> _files = new Queue<string>();
+
+        public FileSystemScanner(string path, bool isDir, DirectoryScannerFactory dirFactory = null)
```

Comment:
- Do we need `isDir` ? shouldn't  `FileSystemScanner` be able to check that?
- `dirFactory` seems to be essential for this thing to work, why is it optional?

Reply:
It could, I had isDir there as the factory would check and then pass that in to set up the scanner properly for if a directory is being checked, but like you said if Scan() is never called then the setup is just unnecessary.

dirFactory is only necessary if we're scanning a directory. If a file is passed in, there's no need to call the "enumerate children" functions; we just return that one file.

I added the DirectoryScanner+Factory in just to cover the possible case of mocking a folder structure without needing to create a temp structure, but if that's not necessary to worry about, this could all pretty much be reverted to the previous setup without DirectoryScanner. The only thing different then would be the added Factory setup to create a scanner.